File: src/python/day04.py
# Input:
example_input = """
..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.
"""


def parse_input(input: str) -> list[list[str]]:
    lines = input.strip().split("\n")
    return [list(line) for line in lines]


def neighbors(x: int, y: int, grid: list[list[str]]) -> list[tuple[int, int]]:
    deltas = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]
    result = []
    for dx, dy in deltas:
        nx, ny = x + dx, y + dy
        if 0 <= nx < len(grid[0]) and 0 <= ny < len(grid):
            result.append((nx, ny))
    return result


def count_rolls(x: int, y: int, grid: list[list[str]]) -> int:
    count = 0
    for nx, ny in neighbors(x, y, grid):
        if grid[ny][nx] == "@":
            count += 1
    return count


def count_total_rolls(grid: list[list[str]]) -> int:
    total = 0
    found = True
    while found == True:
        found = False
        for nx in range(len(grid)):
            for ny in range(0, len(grid[0])):
                if grid[ny][nx] == "@" and count_rolls(nx, ny, grid) < 4:
                    found = True
                    total += 1
                    grid[ny][nx] = "."
    return total


input = open("input/day04.txt", "r").read()
# input = example_input.strip()

m = parse_input(input)
print(count_total_rolls(m))

File: src/rust/day04.rs
use pathfinding::matrix::Matrix;

fn parse_input(input: &str) -> Matrix<char> {
    input.lines().map(|line| line.chars()).collect()
}

fn count_rolls(m: &mut Matrix<char>) -> usize {
    let positions: Vec<_> = m
        .keys()
        .filter(|&pos| m[pos] == '@')
        .filter(|&pos| m.neighbours(pos, true).filter(|&n| m[n] == '@').count() < 4)
        .collect();
    for pos in &positions {
        m[pos] = '.';
    }
    positions.len()
}

pub fn count_rolls_once(input: &str) -> usize {
    let mut m = parse_input(input);
    count_rolls(&mut m)
}

pub fn count_rolls_all(input: &str) -> usize {
    let mut m = parse_input(input);
    let mut total_count = 0;
    loop {
        let count = count_rolls(&mut m);
        total_count += count;
        if count == 0 {
            break;
        }
    }
    total_count
}

pub fn main() {
    let input = std::fs::read_to_string("input/day04.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", count_rolls_once(&input));
    println!("PART 2 = {}", count_rolls_all(&input));
    println!("Execution time: {:?}", now.elapsed());
}
